discord/state.py

```diff
@@ -137,7 +139,200 @@ def done(self) -> None:
+                        deny = Permissions(ow.deny)
+                        overwrite = PermissionOverwrite.from_pair(allow, deny)
+                        if not overwrite.read_messages:
+                            break
+                ret.add(channel)
+
+        return list(ret)
+
+    def parse_args(self, channels: List[Snowflake], force_safe: bool = MISSING) -> Tuple[List[str], List[Tuple[int, int]]]:
+        if not channels:
+            channels = self.get_channels(1 if self.chunk else 5)
+        if force_safe is MISSING:
+            force_safe = self.guild._member_count > 74999
+
+        return [str(channel.id) for channel in channels], self.get_ranges(safe=force_safe)
+
+    def add_members(self, members: List[Member]) -> None:
+        if self.buffer is None:
+            return
+
+        self.buffer.extend(members)
+        if self.cache:
+            guild = self.guild
+            for member in members:
+                guild._add_member(member)
+
+    async def wait(self) -> List[Member]:
+        future = self.loop.create_future()
+        self.waiters.append(future)
+        try:
+            return await future
+        finally:
+            self.waiters.remove(future)
+
+    def get_future(self) -> asyncio.Future[List[Member]]:
+        future = self.loop.create_future()
+        self.waiters.append(future)
+        return future
+
+    def done(self) -> None:
+        for future in self.waiters:
+            if not future.done():
+                future.set_result(self.buffer)
+
+        try:
+            del self.state._scrape_requests[self.guild.id]
+        except KeyError:
+            pass
+
+    def start(self):
+        self.loop.create_task(self.wrapper())
+
+    async def wrapper(self):
+        try:
+            await self.scrape()
+        except RuntimeError:
+            _log.warning('Member list scraping failed for %s.', self.guild.id)
+            self.buffer = None
+        finally:
+            self.done()
+
+    async def scrape(self):
+        self.subscribing = True
+        delay = self.delay
+        channels = self.channels
+        guild = self.guild
+        ws = self.ws
+
+        while self.subscribing:
+            requests = {}
+            for channel in channels:
+                ranges = self.get_current_ranges()
+                if not ranges:
+                    self.subscribing = False
+                    break
+                requests[channel] = ranges
+
+            if not requests:
+                raise RuntimeError('Unable to get channels')
+
+            def predicate(data):
+                if int(data['guild_id']) != guild.id:
+                    return False
+
+                ranges = tuple(chain(requests.values()))
+                ranges_returned = tuple(chain(op.get('ranges', []) for op in data['ops']))
+                return any(range in ranges_returned for range in ranges)
+
+            _log.debug('Subscribing to %s ranges for guild %s.', requests, guild.id)
```

Comment:
You tend to use `%s` commonly over `f-strings`, why so?

Reply:
Because this way the strings aren't resolved unnecessarily if logging is disabled.